src/main/java/org/usfirst/frc4048/common/swervev2/type/GenericSwerveModule.java

```diff
@@ -0,0 +1,65 @@
+package org.usfirst.frc4048.common.swervev2.type;
+
+import edu.wpi.first.math.controller.PIDController;
+import edu.wpi.first.math.controller.ProfiledPIDController;
+import edu.wpi.first.math.controller.SimpleMotorFeedforward;
+import edu.wpi.first.math.geometry.Rotation2d;
+import edu.wpi.first.math.kinematics.SwerveModuleState;
+import edu.wpi.first.math.trajectory.TrapezoidProfile;
+import org.usfirst.frc4048.common.swervev2.components.GenericEncodedSwerve;
+import org.usfirst.frc4048.common.util.Gain;
+import org.usfirst.frc4048.common.util.PID;
+
+public class GenericSwerveModule implements StatedSwerve, PidSwerve {
+    private final PIDController drivePIDController;
+    private final ProfiledPIDController turningPIDController;
+    private final SimpleMotorFeedforward driveFeedforward;
+    private final SimpleMotorFeedforward turnFeedforward;
+    private final GenericEncodedSwerve swerveMotor;
+
+    public GenericSwerveModule(GenericEncodedSwerve swerveMotor, PID drivePid, PID turnPid, Gain driveGain, Gain turnGain, TrapezoidProfile.Constraints goalConstraint) {
+        this.swerveMotor = swerveMotor;
+        drivePIDController = new PIDController(drivePid.getP(),drivePid.getI(),drivePid.getD());
+        turningPIDController = new ProfiledPIDController(turnPid.getP(),turnPid.getI(),turnPid.getD(),goalConstraint);
+        driveFeedforward = new SimpleMotorFeedforward(driveGain.getS(),driveGain.getV());
+        turnFeedforward = new SimpleMotorFeedforward(turnGain.getS(),turnGain.getV());
+        turningPIDController.enableContinuousInput(0, Math.PI * 2);
+    }
+
+    @Override
+    public SwerveModuleState getState() {
+        return new SwerveModuleState(swerveMotor.getDriveEncVel(),new Rotation2d(swerveMotor.getSteerEncPosition()));
+    }
+
+    @Override
+    public void setDesiredState(SwerveModuleState desiredState) {
+        SwerveModuleState state = SwerveModuleState.optimize(desiredState, new Rotation2d(swerveMotor.getSteerEncPosition()));
+        double driveSpeed = calcDrivePidOut(state.speedMetersPerSecond) + calcDriveFeedForward(state.speedMetersPerSecond);
+        double turnSpeed = calcSteerPidOut(state.angle.getRadians()) + calcSteerFeedForward();
+        swerveMotor.getDriveMotor().setVoltage(driveSpeed);
+        swerveMotor.getSteerMotor().set(turnSpeed);
+    }
+
+    @Override
+    public double calcDrivePidOut(double setpoint) {
```

Comment:
This I s overridden from the interface, but should they be externally visible? Maybe they are just internal to the class?

Reply:
Or, maybe this can just be abstract from a superclass of this one?